src/InitControl.cpp

```diff
@@ -0,0 +1,140 @@
+    {
+
+    }
+
+    InitControlImp::InitControlImp(const std::string &input_file,
+                                   PlatformIO &platform_io)
+        : m_platform_io(platform_io)
+    {
+        if (input_file != "") {
+            parse_input(input_file);
+        }
+    }
+
+    void InitControlImp::parse_input(const std::string &input_file)
+    {
+        std::string file_data = read_file(input_file);
+        std::vector<std::string> lines = string_split(file_data, "\n");
+
+        // (      Start capture group
+        // [+-]?  Optional + or -
+        // [0-9]* Zero or more digits
+        // [.]?   Optional .
+        // [0-9]+ One or more digits
+        // (?:    Begin non-capture group
+        // [eE]?  Optional e or E
+        // [+-]?  Optional + or -
+        // [0-9]+ One or more digits
+        // )?     End non-capture group, optional
+        // )      End capture group
+        std::regex sci_notation_regex("([+-]?[0-9]*[.]?[0-9]+(?:[eE]?[+-]?[0-9]+)?)");
+
+        // ^\\s*  String begins with zero or more whitespace characters
+        // (\\w+) Capture group #1 for one or more alphanumeric characters (CONTROL NAME)
+        // \\s+   Zero or more whitespace characters
+        // (\\w+) Capture group #2 for one or more alphanumeric characters (DOMAIN NAME)
+        // \\s+   Zero or more whitespace characters
+        // (\\d+) Capture group #3 for one or more digits (DOMAIN INDEX)
+        // \\s+   Zero or more whitespace characters
+        // (\\S+) Capture group #4 for one or more non-whitespace characters (SETTING)
+        // \\s*   Zero or more whitespace characters
+        std::regex request_regex("^\\s*(\\w+)\\s+(\\w+)\\s+(\\d+)\\s+(\\S+)\\s*");
+        std::smatch match_line;
+
+        for (const auto &line : lines) {
+            // Line parsing is done in 2 phases:
+            //   1. Parse the entire line into match_line with capture groups defined in request_regex.
+            //   2. Parse the SETTING value into match_setting with the capture group defined
+            //      in sci_notation_regex.  Continue to parse it into a double if it is valid.
+            std::regex_search(line, match_line, request_regex);
+
+            if (match_line.empty()) { // No match; possible bad input or comment
+                auto comment_pos = line.find_first_not_of(" \t");
+                if (comment_pos == std::string::npos) {
+                    continue;
+                }
+                else if (line[comment_pos] != '#') {
+                    throw Exception("Invalid line comment or missing fields while parsing: " + line,
+                                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+                }
+            }
+            else { // Match; Continue to try parsing the setting
+                if (match_line.suffix().str().size() > 0 && !string_begins_with(match_line.suffix(), "#")) {
+                    throw Exception("Data found in comment field while parsing: " + line,
```

Comment:
Minor note:

This error message might be a bit confusing. If I'm following correctly, this would get triggered when there's additional text after the portion that was matched, for example:

`CPU_POWER_LIMIT_CONTROL board 0 200   this was supposed to be a comment, but I forgot the #`

The error message would make me think that I included part of the settings within a comment, or that maybe there's something wrong with the domain, domain index or value I provided.

Reply:
Syntax error